util: make ParseMoney return a std::optional<CAmount>

fanquake opened the pull request:
Related discussion in #22193.

fanquake committed:
util: make ParseMoney return a std::optional<CAmount>

MarcoFalke commented:
I was hoping to have both fixes in the same pull in different commits, since they require the reviewers to look up all call sites anyway:
* std::optional
* Check money range inside ParseMoney

theStack commented:
Concept ACK

kiminuo commented:
Concept ACK

_____
Just out of curiosity, I was interested how this changes perf numbers for `ParseMoney`. It's not important though because the function is not used in a repeated way anywhere. Anyway, a [benchmark](c061e79edf04fbf0e05ddbdf4e9a0ec1ba2720f6) says:

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            1,200.00 |          833,333.33 |    1.7% |      0.00 | `MoneyStrParseMoney`
|            1,388.89 |          720,000.00 |    1.8% |      0.00 | `MoneyStrParseMoneyNew`

(No guarrantees, I have tried this for the first time.)

laanwj commented:
Concept ACK, I like this kind of API much better than passing around separate booleans or error codes.

practicalswift commented:
Concept ACK

Non-blocking nit: Personally I think `auto foo = [LHS];` is appropriate only in contexts where the type of `foo` is immediately obvious from reading `[LHS]` (such as when `[LHS]` is `std::make_unique<SomeType>(…)` or `v.begin()`, etc.), or when the type is irrelevant. In the context of this PR I think `std::optional<CAmount> parsed = ParseMoney(…);` would be slightly more clear compared to the suggested `auto parsed = ParseMoney(…);`.

fanquake committed:
util: check MoneyRange() inside ParseMoney()

laanwj commented on src/init.cpp:
```diff
@@ -914,10 +914,12 @@ bool AppInitParameterInteraction(const ArgsManager& args)
-            return InitError(AmountErrMsg("incrementalrelayfee", args.GetArg("-incrementalrelayfee", "")));
-        incrementalRelayFee = CFeeRate(n);
+        const auto arg = args.GetArg("-incrementalrelayfee", "");
+        if (auto inc_relay_fee = ParseMoney(arg)) {
```
TFW no rust-like `if let Some(inc_relay_fee) = Parsemoney(arg) {`

laanwj commented:
> In the context of this PR I think std::optional<CAmount> parsed = ParseMoney(…); would be slightly more clear compared to the suggested auto parsed = ParseMoney(…);.

I hate the verbosity but I do agree in this case, the return type is not obvious.

fanquake commented:
> Check money range inside ParseMoney 

I've added that now.

> I hate the verbosity but I do agree in this case, the return type is not obvious.

Changed to using `std::optional<CAmount>` over `auto`.

I've also rebased and removed some unrelated arg refactoring. Will PR that separately.

MarcoFalke commented on src/init.cpp:
```diff
@@ -965,18 +966,18 @@ bool AppInitParameterInteraction(const ArgsManager& args)
     if (args.IsArgSet("-blockmintxfee")) {
-        CAmount n = 0;
-        if (!ParseMoney(args.GetArg("-blockmintxfee", ""), n))
+        if (!ParseMoney(args.GetArg("-blockmintxfee", "")))
```
```suggestion
        if (!ParseMoney(args.GetArg("-blockmintxfee", ""))) {
```

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -2602,94 +2602,111 @@ std::shared_ptr<CWallet> CWallet::Create(interfaces::Chain* chain, const std::st
-        }
-        if (n > HIGH_TX_FEE_PER_KB) {
+        if (std::optional<CAmount> min_tx_fee = ParseMoney(gArgs.GetArg("-mintxfee", ""))) {
+            if(min_tx_fee.value() == 0) {
```
```suggestion
            if (min_tx_fee.value() == 0) {
```

MarcoFalke replied:
Also wouldn't it be easier to review to keep the `min_tx_fee` symbol outside the if and then keep the if conditions the same?

```cpp
fee = Parse;
if (!fee || !*fee) {
 // same
}
// same

fanquake replied:
Possibly, going to leave for the minute, as I could change this again in a followup.

maflcko reviewed (commented):
approach ACK, didn't review the `src/wallet/wallet.cpp` part yet

fanquake commented:
> approach ACK, didn't review the src/wallet/wallet.cpp part yet

Addressed suggestions so far.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -2602,94 +2602,111 @@ std::shared_ptr<CWallet> CWallet::Create(interfaces::Chain* chain, const std::st
+
+            walletInstance->m_fallback_fee = CFeeRate(fallback_fee.value());
+
+            // Disable fallback fee in case value was set to 0, enable if non-null value
+            walletInstance->m_allow_fallback_fee = walletInstance->m_fallback_fee.GetFeePerK() != 0;
```
Why is this allowed to move inside the `if (gArgs.IsArgSet("-fallbackfee")) {` code block? Prior to this change, `walletInstance->m_fallback_fee` would be set to the default (0), and `walletInstance->m_allow_fallback_fee` would therefore be flipped from its initialized value (true) to false.

fanquake replied:
Think this was just a mistake.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -2602,94 +2602,111 @@ std::shared_ptr<CWallet> CWallet::Create(interfaces::Chain* chain, const std::st
-            return nullptr;
-        }
-        if (n > HIGH_TX_FEE_PER_KB) {
+        if (std::optional<CAmount> min_tx_fee = ParseMoney(gArgs.GetArg("-mintxfee", ""))) {
```
I think you could achieve the same, with a smaller diff and less code duplication as follows:

```diff
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2602,22 +2602,16 @@ std::shared_ptr<CWallet> CWallet::Create(interfaces::Chain* chain, const std::st
     }
 
     if (gArgs.IsArgSet("-mintxfee")) {
-        if (std::optional<CAmount> min_tx_fee = ParseMoney(gArgs.GetArg("-mintxfee", ""))) {
-            if (min_tx_fee.value() == 0) {
-                error = AmountErrMsg("mintxfee", gArgs.GetArg("-mintxfee", ""));
-                return nullptr;
-            }
-
-            if (min_tx_fee.value() > HIGH_TX_FEE_PER_KB) {
-            warnings.push_back(AmountHighWarn("-mintxfee") + Untranslated(" ") +
-                               _("This is the minimum transaction fee you pay on every transaction."));
-            }
-
-            walletInstance->m_min_fee = CFeeRate(min_tx_fee.value());
-        } else {
+        auto min_tx_fee = ParseMoney(gArgs.GetArg("-mintxfee", ""));
+        if (!min_tx_fee || min_tx_fee.value() == 0) {
             error = AmountErrMsg("mintxfee", gArgs.GetArg("-mintxfee", ""));
             return nullptr;
+        } else if (min_tx_fee.value() > HIGH_TX_FEE_PER_KB) {
+            warnings.push_back(AmountHighWarn("-mintxfee") + Untranslated(" ") +
+                               _("This is the minimum transaction fee you pay on every transaction."));
         }
+
+        walletInstance->m_min_fee = CFeeRate(min_tx_fee.value());
     }
 
     if (gArgs.IsArgSet("-maxapsfee")) {
```

MarcoFalke replied:
Suggested the same in https://github.com/bitcoin/bitcoin/pull/22220#discussion_r681491525 ;)

fanquake replied:
Updated to take your suggestions.

jnewbery commented on src/miner.cpp:
```diff
@@ -73,12 +73,8 @@ static BlockAssembler::Options DefaultOptions()
-    } else {
-        options.blockMinFeeRate = CFeeRate(DEFAULT_BLOCK_MIN_TX_FEE);
-    }
+    const CAmount min_tx_fee = ParseMoney(gArgs.GetArg("-blockmintxfee", "")).value_or(DEFAULT_BLOCK_MIN_TX_FEE);
```
I don't think this improves clarity. Previously, `ParseMoney()` would only be called if `-blockmintxfee` is set. Now, we call `ParseMoney()` even if `-blockmintxfee` has not been set, and rely on the fact that it'll return a nullopt if we pass it an empty string.

fanquake replied:
Reverted to something more like the previous code.

MarcoFalke commented:
review ACK f7752adba5dd35fccd3f2144cfcf03538ebf275b 📄

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK f7752adba5dd35fccd3f2144cfcf03538ebf275b 📄
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh9JgwApNfnr4XDam6JFxoS7iOSf7DQeVVVn0nExZiL2GHSj9w0cy83qgzgTGMb
pMxS4Qym8eonS05VJ1lhkFiGWmZ1dyNdxzHw9roxsFw6KXKyzV4SUG9RSKc01fzL
6pNeJ4tC05HuYV6k6MjiI6lRetkWXBf6BYHQFA7QYW1CGi3FVpfavhmd3o5uXr+b
pgb2sWxokm6Gnqs0yq2ZlR3D+hCDxPrgjpj1C0MJ1VC3R4raSwYxetyBrSBFJj/v
oFOFOgVNOhFohA7r+xoehN8TWeoMvywGKxzLGjcGaQSc/HQSKVvtUOeke6mTJBze
jEtSCQlrcxqd+kdVE/+j9NTesaZldMnp84dMEzfbc3dvSz/MThsT3nO/+eG26D9j
wKeynSG82aLymQpH7OqcFho7+FpkqsGjYEwnnpUD5S8EwauQvGoY0Ab/cW2axYMQ
aP/V9Q0PlMiA+rcm/Ut9QFZ4WHhl/eC2FZbUwFW8j6An3sYNRhEietOo+jdc2EN2
cYVt+7du
=cjvz
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `78c354d0f3f17696d32ea10f11f0a9424812baff1f8ecdedfe5fcbc144129d53  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010878c354d0f3f17696d32ea10f11f0a9424812baff1f8ecdedfe5fcbc144129d53f010295b666d7d2b72808d3340136de7dff808fff0106702f6160ea059aa4ffe8305711d930308f10461238ab3f008b1f568f00bdc21580083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010931f6b87bebc93eb3681505b93c26d6308f10461238ab6f00806bab8be20dde6e00083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01096ad5a721f0a4b45b2c6a366f017f02a08f10461238ab5f00865595ed270cc0c9c0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010eb59d43b010a0f19ace8bff57f12df3808f02051598e1c355a51efe55fbcb55ff5f1786b91004c97f20b740782ee067521333d08f10461238ab4f0089d5f5d47238c646f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
